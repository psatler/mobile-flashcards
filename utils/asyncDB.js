import { AsyncStorage } from 'react-native'
import { data } from './mockData'
const FLASHCARDS_DECKS_KEY = 'Flashcards:decks'

const setInitialData = async () => {
    let obj = data;

    await AsyncStorage.setItem(FLASHCARDS_DECKS_KEY, JSON.stringify(obj))
}


export const getDecks = async () => {
    try {
        let value = await AsyncStorage.getItem(FLASHCARDS_DECKS_KEY);
        if( value !== null){ //we have data
            // console.log('decks', value);
            let parsedObj = JSON.parse(value); //parsing back to object
            return parsedObj;
        } else { //if not, we set initial state
            console.log('setting initial state');
            setInitialData(); //defined above
        }
    } catch (err) {
        console.log('There was an error retrieving decks from local DB: ', err);
    }
}

export const saveDeckTitle = (keyTitle, deckObj) => {
    try {
        return AsyncStorage.mergeItem(FLASHCARDS_DECKS_KEY, JSON.stringify({
            [keyTitle]: deckObj
        }))
        
    } catch (err) {
        console.log(err);
    }
}

export const addCardToDeck = (deckTitle, card) => {
    const key = deckTitle;

    return AsyncStorage.getItem(FLASHCARDS_DECKS_KEY)
        .then( (results) => {
            const allDecks = JSON.parse(results);
            const deck = allDecks[key];
            // console.log('addCardToDeck- BEFORE', deck.questions)
            deck.questions.push(card)
            // console.log('addCardToDeck', deck.questions)
            // console.log('allDecks', allDecks)

            AsyncStorage.setItem(FLASHCARDS_DECKS_KEY, JSON.stringify(allDecks)); //updating 

        })
        // const key = deckTitle.split(' ').join('');

}

// 

// const checkIfDataExists = (results) => {
//     return results === null
//       ? setInitialData()
//       : JSON.parse(results)
//   }

// export const retrieveDataFromAsyncStorage = () => {

//     return AsyncStorage.getItem(FLASHCARDS_DECKS_KEY)
//         .then(checkIfDataExists)
// }



// _setValue = async () => {
//     try {
//       await AsyncStorage.setItem(STORAGE_KEY, 'true');
//     } catch (error) { // log the error
//     }
//   };

//   _loadInitialState = async () => {
//     try {
//       let value = await AsyncStorage.getItem(STORAGE_KEY);
//       if (value === 'true'){
//         this.setState({loaded: 'true'});
//       } else {
//         this.setState({loaded: 'false'});
//         this._setValue();
//       }
//     } catch (error) {
//       this.setState({loaded: 'false'});
//       this._setValue();
//     }
//   };