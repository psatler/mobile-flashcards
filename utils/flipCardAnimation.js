import React, { Component } from 'react';
import {
  StyleSheet,
  Text,
  View,
  TouchableOpacity,
  Animated
} from 'react-native';

export default class AnimatedBasic extends Component {

    state = {
        animatedValue: new Animated.Value(0),
        listenerValue: 0,
    }

    componentDidMount(){
        const { animatedValue, listenerValue } = this.state;

        //creating a listener to track values changes
        animatedValue.addListener( ( { value } ) => {
            this.setState({
                listenerValue: value,
            })
        })
    }

    flipCard = () => {
        const { animatedValue, listenerValue } = this.state;

        if(listenerValue >= 90){
            Animated.spring(animatedValue,{
                toValue: 0,
                friction: 8,
                tension: 10
              }).start();
        } 
        else {
            Animated.spring(animatedValue,{
                toValue: 180,
                friction: 8,
                tension: 10
              }).start();
        }
    }

    render() {
        const { animatedValue, listenerValue } = this.state;

        const frontAnimatedStyle = {
          transform: [
            { rotateY: animatedValue.interpolate({
                inputRange: [0, 180],
                outputRange: ['0deg', '180deg'],
              })
            }
          ],
          //adding opacity so the back face won't be mirrored on Android devices
        //https://github.com/facebook/react-native/issues/1973#issuecomment-262059217
          opacity: animatedValue.interpolate({ //adding this so it won't be mirrored on Android devices
                inputRange: [89, 90],
                outputRange: [1, 0]
            }), 
        }
        const backAnimatedStyle = {
          transform: [
            { rotateY: animatedValue.interpolate({
                inputRange: [0, 180],
              //   outputRange: ['180deg', '360deg'],
                outputRange: ['180deg', '0deg']
              }) 
            }
          ],
          opacity: animatedValue.interpolate({  //adding this so it won't be mirrored on Android devices
                inputRange: [89, 90], 
                outputRange: [0, 1] 
            }), 
        }
        
        return (
          <View style={styles.container}>
            <View>
              <Animated.View style={[styles.flipCard, frontAnimatedStyle]}>
                <Text style={styles.flipText}>
                  This text is flipping on the front.
                </Text>
              </Animated.View>
              <Animated.View style={[backAnimatedStyle, styles.flipCard, styles.flipCardBack]}>
                <Text style={styles.flipText}>
                  This text is flipping on the back.
                </Text>
              </Animated.View>
            </View>
            <TouchableOpacity onPress={() => this.flipCard()}>
              <Text>Flip!</Text>
            </TouchableOpacity>
          </View>
        );
      }
    }


const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: "center",
    justifyContent: "center",
  },
  flipCard: {
    width: 200,
    height: 200,
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: 'blue',
    backfaceVisibility: 'hidden', //it means when the card rotates out of the way, it won't be visible. When it rotates back, it'll be visible again. Not working on Android though, so it was added an workaround for this above
  },
  flipCardBack: {
    backgroundColor: "red",
    position: "absolute",
    top: 0,
  },
  flipText: {
    width: 90,
    fontSize: 20,
    color: 'white',
    fontWeight: 'bold',
  }
});