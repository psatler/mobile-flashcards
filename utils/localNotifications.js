// import React from 'react'
import { AsyncStorage } from 'react-native'
import { Notifications, Permissions } from 'expo'

const FLASHCARDS_NOTIFICATIONS_KEY = 'Flashcards:localNotifications'


const createNotification = () => {
    return {
        title: 'Do a quiz today!',
        body: 'ðŸ‘‹ Do not forget to study your decks!',
        ios: {
            sound: true,
        },
        android: {
            sound: true,
            priority: 'high',
            sticky: false,
            vibrate: true,
        }
    }
}

export const setLocalNotification = () => {
    //checking if our local notification has already been set
    AsyncStorage.getItem(FLASHCARDS_NOTIFICATIONS_KEY) 
        .then(JSON.parse) //parsing the returned value
        .then( (data) => {
            if(data === null){ //if we haven't any notification scheduled
                Permissions.askAsync(Permissions.NOTIFICATIONS)
                    .then( ({ status }) => {
                        if( status === 'granted' ){
                            Notifications.cancelAllScheduledNotificationsAsync();

                            let tomorrow = new Date();
                            tomorrow.setDate(tomorrow.getDate() + 1) //tomorrow 
                            tomorrow.setHours(10); //scheduling for 10 AM
                            tomorrow.setMinutes(0);

                            Notifications.scheduleLocalNotificationAsync(
                                createNotification(),
                                {
                                    time: tomorrow,
                                    repeat: 'day'
                                }
                            )

                            AsyncStorage.setItem(FLASHCARDS_NOTIFICATIONS_KEY, JSON.stringify(true))
                        }
                    })

            }
        })

}

export const clearLocalNotification = () => {
    return AsyncStorage.removeItem(FLASHCARDS_NOTIFICATIONS_KEY)
      .then(Notifications.cancelAllScheduledNotificationsAsync)
}