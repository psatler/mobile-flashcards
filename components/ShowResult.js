import React, { Component } from 'react'
import { Text, View, StyleSheet, Platform, TouchableOpacity } from 'react-native'
import { Ionicons, EvilIcons } from '@expo/vector-icons'
import { white, darkBlue, lightBlue } from '../utils/colors';
import { clearLocalNotification, setLocalNotification } from '../utils/localNotifications'
import { withNavigation } from 'react-navigation';

class ShowResult extends Component {

    componentDidMount(){ //when the page mounts, it means the user has completed a deck, so we need to clear the old notification and set a new one
        clearLocalNotification() 
            .then(setLocalNotification); //sets a new local notification for tomorrow 

    }

    render() {
        const { score, singleDeck } = this.props;
        const total = singleDeck.questions.length

        return (
            <View style={styles.container} >
                <View style={{flex: 0.5}}></View> 

                <View style={styles.resultsContainer}>
                    
                    { total === 0 ? ( //there is no cards inside deck
                        <View>
                            <Text style={styles.screenTitle} > There is no cards inside this deck. </Text>
                            <Text style={styles.resultMessage} > Please insert some card and do the quiz again </Text>
                        </View>
                    ) : (
                        <View>
                            <Text style={styles.screenTitle} > Final Result </Text>
                            <Text style={styles.resultMessage} > You've Got {score} out of {total} points possible! </Text>
                        </View>
                    )
                }

                    <View style={styles.buttonsContainer} >

                        { total !== 0 ? (
                            <TouchableOpacity 
                            style={ {alignItems: 'center' }}  
                            onPress={ () => this.props.restartFunc()}    
                            >
                                <EvilIcons name={'redo' } size={50} color={lightBlue}  />
                                <Text style={{color: lightBlue}}> Restart Quiz </Text>
                            </TouchableOpacity>
                        ) : ( null ) }

                        <TouchableOpacity 
                            style={ { 
                                alignItems: 'center', 
                                } }
                            onPress={ () => this.props.navigation.navigate('DeckDetail', {
                                deckTitle: singleDeck.title,
                                deckLength: singleDeck.questions.length
                            })}        
                        >
                            <Ionicons name={Platform.OS === 'ios' ? 'ios-home' : 'md-home' } size={50} color={lightBlue}  />
                            <Text style={{color: lightBlue}}> Back to Deck </Text>
                        </TouchableOpacity>
                    </View>
            
                </View>

                <View style={{flex: 1}}></View>
            </View>
        )
    }
}

const styles = StyleSheet.create({
    container: {
        flex: 1,
    },
    resultsContainer: {
        flex: 2, //to make it fill the majority of the space available
        margin: 10,
        borderWidth: 1,
        borderColor: 'black',
        borderRadius: Platform.OS === 'ios' ? 16 : 2,
        backgroundColor: darkBlue,
    },
    screenTitle: {
        fontSize: 30,
        textAlign: 'center',
        marginTop: 20,
        marginBottom: 30,
        color: lightBlue,
    },
    resultMessage: {
        fontSize: 15,
        textAlign: 'center',
        color: white,
    },
    buttonsContainer: {
        flexDirection: 'row',
        justifyContent: 'space-evenly' ,
        alignItems: 'center',
        marginTop: 30,
    }


})

export default withNavigation(ShowResult); //using withNavigation HoC so we can access the navigation props