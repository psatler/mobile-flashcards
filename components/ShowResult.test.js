import React from 'react'
import { ShowResult } from './ShowResult'

import renderer from 'react-test-renderer';
import ShallowRenderer from 'react-test-renderer/shallow';

import { data } from '../utils/mockData'


describe('[Component] Deck', () => {
    const restartQuiz = jest.fn();
    const deck = data['JavaScript']; //javascript deck
    const deckNoCards = data['React'];
    deckNoCards.questions = []
    // const deckWithNoCards = {"newDeck": {"questions": [], "title": "new Deck"}} 

    it('should show a message saying the score and two buttons: restart quiz and back to deck', () => {
        const score = 0;
        const rendered = renderer.create(
            <ShowResult score={score} singleDeck={deck} restartFunc={restartQuiz} />
        ).toJSON();
        expect(rendered).toMatchSnapshot();
    });

    it('should show a message saying there is no cards inside deck and display only one button: back to deck', () => {
        const score = 0;
        const renderer = new ShallowRenderer();
        renderer.render(
            <ShowResult score={score} singleDeck={deckNoCards} restartFunc={restartQuiz} />
        );
        const tree = renderer.getRenderOutput();
        expect(tree).toMatchSnapshot();
    });

});