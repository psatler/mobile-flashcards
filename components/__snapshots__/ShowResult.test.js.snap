// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Component] Deck should show a message saying the score and two buttons: restart quiz and back to deck 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flex": 0.5,
      }
    }
  />
  <View
    style={
      Object {
        "backgroundColor": "#18263D",
        "borderColor": "black",
        "borderRadius": 16,
        "borderWidth": 1,
        "flex": 2,
        "margin": 10,
      }
    }
  >
    <View>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#00A4D3",
            "fontSize": 30,
            "marginBottom": 30,
            "marginTop": 20,
            "textAlign": "center",
          }
        }
      >
         Final Result 
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#fff",
            "fontSize": 15,
            "textAlign": "center",
          }
        }
      >
         You've Got 
        0
         out of 
        1
         points possible! 
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-evenly",
          "marginTop": 30,
        }
      }
    >
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        collapsable={undefined}
        hasTVPreferredFocus={undefined}
        hitSlop={undefined}
        isTVSelectable={true}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "opacity": 1,
          }
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        />
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#00A4D3",
            }
          }
        >
           Restart Quiz 
        </Text>
      </View>
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        collapsable={undefined}
        hasTVPreferredFocus={undefined}
        hitSlop={undefined}
        isTVSelectable={true}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "opacity": 1,
          }
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
        />
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#00A4D3",
            }
          }
        >
           Back to Deck 
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  />
</View>
`;

exports[`[Component] Deck should show a message saying there is no cards inside deck and display only one button: back to deck 1`] = `
<Component
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Component
    style={
      Object {
        "flex": 0.5,
      }
    }
  />
  <Component
    style={
      Object {
        "backgroundColor": "#18263D",
        "borderColor": "black",
        "borderRadius": 16,
        "borderWidth": 1,
        "flex": 2,
        "margin": 10,
      }
    }
  >
    <Component>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#00A4D3",
            "fontSize": 30,
            "marginBottom": 30,
            "marginTop": 20,
            "textAlign": "center",
          }
        }
      >
         There is no cards inside this deck. 
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#fff",
            "fontSize": 15,
            "textAlign": "center",
          }
        }
      >
         Please insert some card and do the quiz again 
      </Text>
    </Component>
    <Component
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-evenly",
          "marginTop": 30,
        }
      }
    >
      <TouchableOpacity
        activeOpacity={0.2}
        onPress={[Function]}
        style={
          Object {
            "alignItems": "center",
          }
        }
      >
        <Icon
          allowFontScaling={false}
          color="#00A4D3"
          name="ios-home"
          size={50}
        />
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "color": "#00A4D3",
            }
          }
        >
           Back to Deck 
        </Text>
      </TouchableOpacity>
    </Component>
  </Component>
  <Component
    style={
      Object {
        "flex": 1,
      }
    }
  />
</Component>
`;
